tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -1142,26 +1173,47 @@ public:
    SumHelper(SumHelper &&) = default;
    SumHelper(const SumHelper &) = delete;
    SumHelper(const std::shared_ptr<ResultType> &sumVPtr, const unsigned int nSlots)
-      : fResultSum(sumVPtr), fSums(nSlots, NeutralElement(*sumVPtr, -1))
+      : fResultSum(sumVPtr), fSums(nSlots, NeutralElement(*sumVPtr, -1)),
+        fCompensations(nSlots, NeutralElement(*sumVPtr, -1))
    {
    }
-
    void InitTask(TTreeReader *, unsigned int) {}
-   void Exec(unsigned int slot, ResultType v) { fSums[slot] += v; }
+
+   void Exec(unsigned int slot, ResultType x)
+   {
+      // Kahan Sum:
+      ResultType y = x - fCompensations[slot];
+      ResultType t = fSums[slot] + y;
+      fCompensations[slot] = (t - fSums[slot]) - y;
+      fSums[slot] = t;
+   }
 
    template <typename T, std::enable_if_t<IsDataContainer<T>::value, int> = 0>
    void Exec(unsigned int slot, const T &vs)
    {
-      for (auto &&v : vs)
-         fSums[slot] += static_cast<ResultType>(v);
+      for (auto &&v : vs) {
+         Exec(slot, v);
+      }
    }
 
-   void Initialize() { /* noop */}
+   void Initialize()
+   { /* noop */
+   }
 
    void Finalize()
    {
-      for (auto &m : fSums)
-         *fResultSum += m;
+      ResultType sum(NeutralElement(ResultType{}, -1));
+      ResultType compensation(NeutralElement(ResultType{}, -1));
+      ResultType y(NeutralElement(ResultType{}, -1));
+      ResultType t(NeutralElement(ResultType{}, -1));
+      for (auto &m : fSums) {
+         // Kahan Sum:
+         y = m - compensation;
+         t = sum + y;
+         compensation = (t - sum) - y;
```

Comment:
shouldn't this take a vector of compensations?

Reply:
No, because this is just the Kahan sum of the final contributions which are already sums and not vectors.